pkg/code-generator/schemagen/crd.go

```diff
@@ -0,0 +1,123 @@
+package schemagen
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/rotisserie/eris"
+
+	code_generator "github.com/solo-io/solo-kit/pkg/code-generator"
+	"github.com/solo-io/solo-kit/pkg/code-generator/writer"
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/utils/pointer"
+
+	"github.com/ghodss/yaml"
+	kubeyaml "k8s.io/apimachinery/pkg/util/yaml"
+)
+
+const (
+	apiVersion = "apiextensions.k8s.io/v1beta1"
+)
+
+var (
+	ApiVersionMismatch = func(expected, actual string) error {
+		return eris.Errorf("Expected ApiVersion [%s] but found [%s]", expected, actual)
+	}
+)
+
+type CrdWriter struct {
+	fileWriter writer.FileWriter
+}
+
+func NewCrdWriter(crdDirectory string) *CrdWriter {
+	return &CrdWriter{
+		fileWriter: &writer.DefaultFileWriter{
+			Root: crdDirectory,
+			HeaderFromFilename: func(s string) string {
+				return fmt.Sprintf("# %s\n\n", writer.DefaultFileHeader)
+			},
+		},
+	}
+}
+
+func (c *CrdWriter) ApplyValidationSchemaToCRD(crd v1beta1.CustomResourceDefinition, validationSchema *v1beta1.CustomResourceValidation) error {
+	crd.Spec.Validation = validationSchema
+	// Setting PreserveUnknownFields to false ensures that objects with unknown fields are rejected.
+	// This is deprecated and will default to false in future versions.
+	crd.Spec.PreserveUnknownFields = pointer.BoolPtr(false)
+
+	crdBytes, err := yaml.Marshal(crd)
+	if err != nil {
+		return err
+	}
+
+	return c.fileWriter.WriteFile(code_generator.File{
+		Filename: c.getFilenameForCRD(crd),
+		Content:  string(crdBytes),
+	})
+}
+
+func (c *CrdWriter) getFilenameForCRD(crd v1beta1.CustomResourceDefinition) string {
+	return fmt.Sprintf("%s_%s.yaml", crd.Spec.Group, crd.Spec.Names.Kind)
```

Comment:
worth including the version to get the entire GVK in the filename?

Reply:
Yeah I had opted out of including the version, since many versions can be defined on the CRD itself. However, doesn't hurt to add it, can always rename later.